ddtrace/contrib/anthropic/utils.py

```diff
@@ -0,0 +1,284 @@
+import sys
+from typing import AsyncGenerator
+from typing import Dict
+from typing import Generator
+from typing import Tuple
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.vendor import wrapt
+
+
+log = get_logger(__name__)
+
+
+def handle_stream_response(integration, resp, args, kwargs, span):
+    if _is_async_generator(resp):
+        return TracedOpenAIAsyncStream(resp, integration, span, args, kwargs)
+    elif _is_generator(resp):
+        return TracedOpenAIStream(resp, integration, span, args, kwargs)
+    elif _is_stream_manager(resp):
+        return TracedOpenAIStreamManager(resp, integration, span, args, kwargs)
+
+
+def _process_finished_stream(integration, span, args, kwargs, streamed_chunks, message_accumulated=False):
+    messages = None
+    messages = kwargs.get("messages", None)
+    chat_messages = get_argument_value(args, kwargs, 0, "messages")
+    try:
+        if message_accumulated:
+            messages = _construct_accumulated_messages(streamed_chunks)
+        else:
+            messages = _construct_unaccumulated_messages(streamed_chunks)
+        if integration.is_pc_sampled_span(span):
+            _tag_streamed_chat_completion_response(integration, span, messages)
+        if integration.is_pc_sampled_llmobs(span):
+            integration.llmobs_set_tags(
+                span,
+                input_messages=chat_messages,
+                formatted_response=messages,
+            )
+    except Exception:
+        log.warning("Error processing streamed completion/chat response.", exc_info=True)
+
+
+class BaseTracedOpenAIStream(wrapt.ObjectProxy):
+    def __init__(self, wrapped, integration, span, args, kwargs, is_message_stream=False):
+        super().__init__(wrapped)
+        n = kwargs.get("n", 1) or 1
+        self._dd_span = span
+        self._streamed_chunks = [[] for _ in range(n)]
+        self._dd_integration = integration
+        self._kwargs = kwargs
+        self._args = args
+        self._is_message_stream = (
+            is_message_stream  # message stream helper will have accumulated the message for us already
+        )
```

Comment:
I'm a bit confused on what this does. Could you elaborate what's happening here?

Reply:
Trying to find a better name. When calling `messages.stream()` as opposed to `messages.create(Stream=true)`, Anthropic uses a `messages.Stream` helper class that will have already accumulated the streamed message within the first stream content chunk. It means we don't have to iterate over every stream chunk to build up the full response message.

Reply:
Removed this code